innkeeper/client.go

```diff
@@ -0,0 +1,417 @@
+type FixedToken string
+
+func (fa FixedToken) Token() (string, error) { return string(fa), nil }
+
+type pathMatch struct {
+	Typ   pathMatchType `json:"type"`
+	Match string        `json:"match"`
+}
+
+type pathRewrite struct {
+	Match   string `json:"match"`
+	Replace string `json:"replace"`
+}
+
+type headerData struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+type endpoint struct {
+	Typ      endpointType `json:"type"`
+	Protocol string       `json:"protocol"`
+	Hostname string       `json:"hostname"`
+	Port     int          `json:"port"`
+	Path     string       `json:"path"`
+}
+
+type routeDef struct {
+	MatchMethods    []string     `json:"match_methods"`
+	MatchHeaders    []headerData `json:"match_headers"`
+	MatchPath       pathMatch    `json:"match_path"`
+	RewritePath     *pathRewrite `json:"rewrite_path"`
+	RequestHeaders  []headerData `json:"request_headers"`
+	ResponseHeaders []headerData `json:"response_headers"`
+	Endpoint        endpoint     `json:"endpoint"`
+}
+
+type routeData struct {
+	Id        int64    `json:"id"`
+	CreatedAt string   `json:"createdAt"`
+	DeletedAt string   `json:"deletedAt"`
+	Route     routeDef `json:"route"`
+}
+
+type apiError struct {
+	Message   string `json:"message"`
+	ErrorType string `json:"type"`
+}
+
+type routeDoc string
+
+type Client struct {
+	baseUrl      *url.URL
+	auth         Authentication
+	authToken    string
+	lastModified string
+	httpClient   *http.Client
+	dataChan     chan skipper.RawData
+
+	// store the routes for comparison during
+	// the subsequent polls
+	doc    map[int64]string
+	closer chan interface{}
+	closed chan interface{}
+}
+
+type Options struct {
+	Address        string
+	Insecure       bool
+	PollTimeout    time.Duration
+	Authentication Authentication
+}
+
+// Creates an innkeeper client.
+func Make(o Options) (*Client, error) {
+	u, err := url.Parse(o.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		u, o.Authentication, "", "",
+		&http.Client{Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: o.Insecure}}},
+		make(chan skipper.RawData),
+		make(map[int64]string),
+		make(chan interface{}),
+		make(chan interface{})}
+
+	// start a polling loop
+	go func() {
+		to := time.Duration(0)
+		for {
+			select {
+			case <-time.After(to):
+				to = o.PollTimeout
+				if c.poll() {
+					c.dataChan <- toDocument(c.doc)
+				}
+			case <-c.closer:
+				return
+			}
+		}
+	}()
+
+	return c, nil
+}
+
+func (c *Client) createUrl(path ...string) string {
+	u := *c.baseUrl
+	u.Path = strings.Join(append([]string{""}, path...), "/")
+	return (&u).String()
+}
+
+func parseApiError(r io.Reader) (string, error) {
+	message, err := ioutil.ReadAll(r)
+
+	if err != nil {
+		return "", err
+	}
+
+	ae := apiError{}
+	if err := json.Unmarshal(message, &ae); err != nil {
+		return "", err
+	}
+
+	return ae.ErrorType, nil
+}
+
+func isApiAuthError(error string) bool {
+	switch authErrorType(error) {
```

Comment:
A switch to either return true or false doesn't feel right... why not somethin like `return error == a || error == b`?